Move Shared Mem Transpose Detection Logic to LinalgOpTrait

KoolJBlack opened the pull request:
Cleanup to consolidate code and improve readability. 

Related to #10005

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -91,6 +92,40 @@ static bool isTransposeLinalgOp(linalg::LinalgOp linalgOp) {
 
+/// Returns true if the index map represents a transpose that benefits from
+/// shared mem.
+static bool isSharedMemTranspose(AffineMap indexMap) {
```
The shared memory naming here and below doesn't really make sense as part of a generic matcher. Shared memory is the implementation used for the GPU codegen but what we are trying to detect is whether we are transposing the most inner dimension. I think we should pick better naming to describe that part and separate it from the implementation.

dcaballe commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -91,6 +92,40 @@ static bool isTransposeLinalgOp(linalg::LinalgOp linalgOp) {
+    return false;
+  }
+  return true;
+}
```
These also look like specific constraints for your case as a transpose can have more than 3 dimensions, dynamic shapes, reduction dimensions, etc. We can add another `[](LinalgOp op){}` lambda, to filter out LinalgOps, in this case.

dcaballe commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -91,6 +92,40 @@ static bool isTransposeLinalgOp(linalg::LinalgOp linalgOp) {
+        indexMap.getNumDims() - 1) {
+      return true;
+    }
+  }
```
IIUC, we need to filter out projected permutations and transposes that do not permute the last dimension. I think those are very specific constraints that we couldn't reuse in general. However, we can extend `isTranspose` and pass a `[](AffineMap){...}` lambda that applies specific constrains to the map. In that way, the main logic would be generic and reusable but we could also extending it with specific checks. WDYT?

dcaballe commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -99,9 +134,40 @@ static bool computeReductionInfo(LinalgOp linalgOp) {
+      return sharedMemTransposeOperands;
+    }
+  }
+
```
Same here. I think all these checks above should be part of the LinalgOp lambda function.

dcaballe commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -99,9 +134,40 @@ static bool computeReductionInfo(LinalgOp linalgOp) {
+    }
+  }
+
+  // Determine which operands are transposed.
```
These is great! We could move this to `isTranspose` and have just one implementation.

dcaballe reviewed (changes requested):
Awesome! Thanks for the refactoring!

I think in general the idea is to keep in `LinalgOpInfo` only the code that is generic enough to be reusable. We can extend this class to take a couple of lambda filters, one for AffineMap and another one for LinalgOp, to pass specific constraints from each client that are not generic enough to be reusable but should be used to make the transpose property more specific. I'm adding more comments inline. Happy to quickly discuss this over GVC!

KoolJBlack commented:
> 

@dcaballe on closer inspection, I've realized the `isTransposeLinalgOp` function does quite a few things including using different inverses for the input/output operands this will prolly be much harder to generalize
perhaps its not worth trying to merge these two codes, unless you know if what can be made generic in this logic
it seems its already detecting a particular type of transpose specific to CPU?

If I'm bypassing all of that logic, there is very little I can reuse on the GPU side that isn't just single function calls. 

I'm now thinking it might make more sense to consolidate the transpose code in a LLVMGPU specific shared header that can be reused throughout.

dcaballe commented:
> @dcaballe on closer inspection, I've realized the `isTransposeLinalgOp` function does quite a few things including using different inverses for the input/output operands this will prolly be much harder to generalize perhaps its not worth trying to merge these two codes, unless you know if what can be made generic in this logic it seems its already detecting a particular type of transpose specific to CPU?

This logic is not detecting anything specific to CPUs or any specific transposes. It's just detecting any transpose operation. There are, of course, some TODOs that will be incrementally addressed but it's nothing specific. The map inversion is due to the indexing map semantics in a LinalgOp. They represent how to transform the iteration space indexes for a specific input/output. However, to actually model the memory transformations that will happen out of those indexing maps, we have to transform them accordingly, hence the inversions that you see. Again, this is nothing specific. The inversion and related code is what we use to generate the actual transfer ops and transpose operations out of the generic op.
 
> If I'm bypassing all of that logic, there is very little I can reuse on the GPU side that isn't just single function calls.

That's why I mentioned that the filter should add extra constraints, not just replace the existing one. The existing one is generic and will match all the transposes (eventually). An extra filter should build on top of the generic one and add more constraints to narrow down the output.

> I'm now thinking it might make more sense to consolidate the transpose code in a LLVMGPU specific shared header that can be reused throughout.

I guess that's up to you but as you can see the logic will get complicated the moment to try to match more cases than the current one for the FVD.

KoolJBlack commented:
I've reworked the transpose info to support some more general traits that are useful for GPU. I've also removed the specific shared mem transpose code and added to a separate header. This is supplied as an additional filter for transposed ops.

dcaballe commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -57,51 +56,84 @@ static bool isTransposeMap(AffineMap map) {
+  computeInfo(linalgOp);
+}
+LinalgOpInfo::LinalgOpInfo(linalg::LinalgOp linalgOp,
+                           TransposeMapFilter transposeMapFilter)
```
Great! (Nothing to address as part of this PR) Eventually this should move to a setter because otherwise this won't scale if we need more and more filters. Also, we shouldn't compute all the information at construction time.

KoolJBlack replied:
agreed.

dcaballe commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -57,51 +56,84 @@ static bool isTransposeMap(AffineMap map) {
+      return transposeOperands;
+    }
+    AffineMap inverseMap = inverseAndBroadcastProjectedPermutation(map);
+    if (isTransposeMap(inverseMap) && transposeMapFilter(inverseMap)) {
```
awesome! thanks!

dcaballe commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -57,51 +56,84 @@ static bool isTransposeMap(AffineMap map) {
 
   // Inverse map to use transfer op permutation logic.
   AffineMap outputInversedMap = inversePermutation(
       linalgOp.getTiedIndexingMap(linalgOp.getOutputOperand(0)));
```
Would you mind moving this next to their use in line 106?

dcaballe commented on compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp:
```diff
@@ -57,51 +56,84 @@ static bool isTransposeMap(AffineMap map) {
+
+static bool computeGenericInfo(LinalgOp linalgOp) {
+  return isa<GenericOp>(linalgOp);
+}
```
Would it make more sense to do something like:
```
if (linalgOp.hasDynamicShape() || hasTwoOrThreeLoopsInfo(linalgOp) || isa<GenericOp>(linalgOp))
  return; // or bail out

LinalgOpInfo opInfo (linalgOp);
```

and keep these checks/traits out of LinalgOpInfo? That information is easy to get from the linalgOp itself. I think it's good if we use linalgOp directly when possible instead of adding a wrapper around a simple check in LinalgOpInfo. Otherwise, LinalgOpInfo will grow a lot and will duplicate queries that are already in LinalgOp. WDYT?

I think we should also remove the reductionTrait. I can take care of that later.

KoolJBlack replied:
Done. Though I left the `isReduction` trait. I think that is general enough that it will be useful around but you can remove it if you choose.

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMGPU/KernelConfig.cpp:
```diff
@@ -511,77 +513,27 @@ static LogicalResult setWarpReductionConfig(func::FuncOp entryPoint,
-    }
+  // Checks preconditions for shared mem transpose.
+  if (!opInfo.isTranspose() || opInfo.isDynamic() || opInfo.isReduction() ||
+      !opInfo.isTwoThreeLoops()) {
```
yeah, I would use linalgOp for the simple checks here and move the `isTwoThreeLoops` check to this file since it's a specific check.

dcaballe reviewed (commented):
Looks great! Just a couple of comments. Thanks!

dcaballe reviewed (approved):
Thanks! LGTM!
